### ISSUE TITLE
Cannot borrow as mutable despite DerefMut

### ISSUE THREAD

**Issue by smcmurray** [rust-lang/rust#51886]
May be related to #17501

````rust
use std::ops::{Deref, DerefMut};

struct Test(Box<FnMut()>);

impl Deref for Test {
    type Target = Box<FnMut()>;
    fn deref(&self)->&Self::Target {
        &self.0
    }
}

impl DerefMut for Test {
    fn deref_mut(&mut self)-> &mut Box<FnMut()> {
        &mut self.0
    }
}

fn main() {
    let mut t = Test(Box::new(move || ()));
    (t.0)();
    t();
}
````

Rust 2015:

```
error[E0596]: cannot borrow immutable `Box` content as mutable
  --> src/main.rs:21:5
   |
21 |     t();
   |     ^ cannot borrow as mutable
```

Rust 2018:

```
error[E0596]: cannot borrow data in a `&` reference as mutable
  --> src/main.rs:21:5
   |
21 |     t();
   |     ^ cannot borrow as mutable
```

**Comment by cuviper** [comment]
FWIW, I tried to hide the `Box` aspect, but it also fails with `type Target = dyn FnMut()`.

**Comment by kuviman** [comment]
It looks like anything but `Box<T: FnMut()>` and `T: FnMut()` itself, cause this error:

```rust
fn foo<F: FnMut(), W: Deref<Target = F> + DerefMut>(mut f: W) {
    f()
}
```

```
warning: variable does not need to be mutable
 --> src/lib.rs:3:53
  |
3 | fn foo<F: FnMut(), W: Deref<Target = F> + DerefMut>(mut f: W) {
  |                                                     ----^
  |                                                     |
  |                                                     help: remove this `mut`
  |
  = note: #[warn(unused_mut)] on by default

error[E0596]: cannot borrow data in a `&` reference as mutable
 --> src/lib.rs:4:5
  |
4 |     f()
  |     ^ cannot borrow as mutable
```

Based on the warning, it looks like only `Deref` is being considered, and `DerefMut` is not

**Comment by cuviper** [comment]
It's happy if you force `DerefMut` like this: `(&mut *t)()`

**Comment by rylev** [comment]
Triage: [This no longer seems to be an issue](https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=43d36c949903a357ab9c9c0b42bdc6fc) as both code samples from above compile successfully now. This seems to have been fixed in 1.46.
